### DISCUSSION TITLE
ky-universal throws error "Cannot use import statement outside a module"

### DISCUSSION THREAD

**Discussion by oldo** [vercel/next.js#21252] [Help]
**What version of Next.js are you using?**

10.0.5

**What version of Node.js are you using?**

12.13.0

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS

**How are you deploying your application?**

next start

**Describe the Bug**

I would like to use [ky-universal](https://github.com/sindresorhus/ky-universal) in my application to make http requests, however just simple usage throws the error:

```
import fetch, {Headers, Request, Response} from 'node-fetch';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

I've tried creating an issue on `ky-universal` repo and I was brushed off and told to ask here instead.

**Expected Behavior**

No error should be thrown and request should be successfully made.

**To Reproduce**

I just have this in a `useEffect` hook in one of my component:

```
useEffect(() => {
    const response = ky.get('https://test.com');
    console.log(response);
  });
```

**Comment by kvnwolf** [comment]
This is happening because you can only consume _transpiled_ packages on Next.js. Since Next.js is built on top of Webpack, you can use [`next-transpile-modules`](https://www.npmjs.com/package/next-transpile-modules) to [extend the default Webpack configuration](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config) and use the [`babel-loader`](https://www.npmjs.com/package/babel-loader) on your input files, applying the internal [`nextjs preset`](https://github.com/vercel/next.js/blob/canary/packages/next/build/babel/preset.ts) to the defined external modules.

#### 1. Install `next-transpile-modules`

```sh
yarn add --D next-transpile-modules
```

#### 2. Require `next-transpile-modules`, create a `withTM` function passing in the modules you need to be transpiled as an argument, and export your config wrapped `withTM` in `next.config.js`

```js
const transpileModules = require("next-transpile-modules")

const withTM = transpileModules(["ky-universal"])

module.exports = withTM()
```

**Reply by reducio** [reply]
Thanks for answer, but don't work for `ky/universal`.

**Reply by sholladay** [reply]
To the Next.js team: `ky-universal` uses modern ESM syntax and top-level await but we work natively in Node.js _and_ browsers _without_ transpiling. We need a clean way to tell Next.js users how to prevent unnecessarily transpiling us because Next.js (or at least Webpack) seems to break our usage of top-level await - which, again, does not need to be transpiled to begin with in most environments.

Also, this should be reported as an issue instead of a discussion, IMO.

**Reply by reducio** [reply]
Yes, you are right. I tried something like this, but top-level await problem is not solved:

```javascript
// next.config.js

const transpileModules = require('next-transpile-modules')

const withTM = transpileModules(['ky-universal'], { unstable_webpack5: true })

module.exports = withTM({
    env: {
        ...
    },
    webpack: (config, _options) => {
        config.experiments = {
            topLevelAwait: true,
            importAsync: true,
            asyncWebAssembly: true,
            importAwait: true,
        }
        return config
    },
})
```

**Reply by TheGooner93** [reply]
Do we have any updates on this ? `ky-universal` is pretty unusable with NextJS due to this conundrum.

**Reply by sholladay** [reply]
I don't think they are going to notice in the Discussions tab.

**Reply by reducio** [reply]
> Do we have any updates on this ? `ky-universal` is pretty unusable with NextJS due to this conundrum.

Yesterday, I updated project dependencies on base Next.js and ky, ky-universal, everything works fine.

"next": "^10.2.0",
"ky": "^0.28.0",
"ky-universal": "^0.9.1",

**Reply by gaurav5430** [reply]
>This is happening because you can only consume transpiled packages on Next.js

what do you mean it can only consume transpiled packages ?

**Reply by nghiepdev** [reply]
> Yesterday, I updated project dependencies on base Next.js and ky, ky-universal, everything works fine.

I just tried the update and it didn't work.

**Reply by LeonAlvarez** [reply]
Just tried it but im getting the following error
```error - ./node_modules/ky-universal/index.js
Module parse failed: The top-level-await experiment is not enabled (set experiments.topLevelAwait: true to enabled it)
File was processed with these loaders:
 * ./node_modules/next/dist/build/webpack/loaders/next-babel-loader.js
You may need an additional loader to handle the result of these loaders.
Error: The top-level-await experiment is not enabled (set experiments.topLevelAwait: true to enabled it)
```
